File: find-largest-value-in-each-tree-row/flvietr.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func largestValues(root *TreeNode) []int {
	var next []*TreeNode = []*TreeNode{root}
	var vv int
	re := []int{}
	for {
		next, vv = helper(next)
		if len(next) == 0 {
			break
		}
		re = append(re, vv)
		fmt.Println(re)
	}
	return re
}

func helper(root []*TreeNode) ([]*TreeNode, int) {
	var max int = -1 << 31
	nextRound := []*TreeNode{}
	for _, v := range root {
		if v == nil {
			continue
		}

		if v.Val > max {
			max = v.Val
		}
		nextRound = append(nextRound, v.Left, v.Right)
	}

	return nextRound, max
}

func main() {
	a := TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val:   3,
			Left:  &TreeNode{Val: 5},
			Right: &TreeNode{Val: 3},
		},
		Right: &TreeNode{
			Val:   2,
			Right: &TreeNode{Val: 9},
		},
	}

	fmt.Println(largestValues(&a))

	a = TreeNode{
		Val: 88,
		Left: &TreeNode{
			Val:   -44,
			Left:  nil,
			Right: &TreeNode{Val: 56},
		},
		Right: nil,
	}

	fmt.Println(largestValues(&a))
}

File: find-largest-value-in-each-tree-row/flvietr.rs
use std::{cell::RefCell, rc::Rc};

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn largest_values(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let mut re = vec![];
    if root.is_none() {
        return re;
    }

    let mut next = vec![root.unwrap()];
    let mut vv: i32;

    loop {
        (next, vv) = helper(&next);
        if next.len() == 0 {
            re.push(vv);
            break;
        }

        re.push(vv);
    }
    re
}

fn helper(trees: &[Rc<RefCell<TreeNode>>]) -> (Vec<Rc<RefCell<TreeNode>>>, i32) {
    let mut max = i32::MIN;
    let mut next_round = vec![];

    for v in trees {
        if v.borrow().val > max {
            max = v.borrow().val.clone()
        }

        match (&v.borrow().left, &v.borrow().right) {
            (None, None) => (),
            (None, Some(r)) => next_round.push(r.clone()),
            (Some(l), None) => next_round.push(l.clone()),
            (Some(l), Some(r)) => {
                next_round.push(r.clone());
                next_round.push(l.clone());
            }
        }
    }

    (next_round, max)
}

fn main() {}
